From 2311f03cad64ae87bde0985163b3c5f4f4a0fb43 Mon Sep 17 00:00:00 2001
From: James Morse <james.morse@arm.com>
Date: Mon, 22 Jan 2024 14:58:24 +0800
Subject: [PATCH 672/960] downstream: perf/arm-cmn: Stop claiming all the
 resources

Carl reports that when both the MPAM driver and CMN driver are built
into the kernel, they fight over who can claim the resources associated
with their registers. This prevents the second of these two drivers
from probing.

Currently the CMN PMU driver claims all the CMN registers.

The MPAM registers are grouped together in a small number of pages,
whereas the PMU registers that the CMN PMU driver uses appear
throughout the CMN register space.

Having the CMN driver claim all the resources is the wrong thing
to do, and claiming individual registers here and there is not
worthwhile.
Instead, stop the CMN driver from claiming any resources as its
registers are not grouped together.

Reported-by: Carl Worth <carl@os.amperecomputing.com>
Tested-by: Carl Worth <carl@os.amperecomputing.com>
Signed-off-by: James Morse <james.morse@arm.com>
CC: Ilkka Koskinen <ilkka@os.amperecomputing.com>
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/perf/arm-cmn.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/perf/arm-cmn.c b/drivers/perf/arm-cmn.c
index 77aa37de5988..2f022d1fb052 100644
--- a/drivers/perf/arm-cmn.c
+++ b/drivers/perf/arm-cmn.c
@@ -2435,6 +2435,7 @@ static int arm_cmn_probe(struct platform_device *pdev)
 	struct arm_cmn *cmn;
 	const char *name;
 	static atomic_t id;
+	struct resource *cfg;
 	int err, rootnode, this_id;
 
 	cmn = devm_kzalloc(&pdev->dev, sizeof(*cmn), GFP_KERNEL);
@@ -2450,7 +2451,16 @@ static int arm_cmn_probe(struct platform_device *pdev)
 		rootnode = arm_cmn600_acpi_probe(pdev, cmn);
 	} else {
 		rootnode = 0;
-		cmn->base = devm_platform_ioremap_resource(pdev, 0);
+
+		/*
+		 * Avoid registering resources as the PMUs registers are
+		 * scattered through CMN, and may appear either side of
+		 * registers for other 'devices'. (e.g. the MPAM MSC controls).
+		 */
+		cfg = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+		if (!cfg)
+			return -EINVAL;
+		cmn->base = devm_ioremap(&pdev->dev, cfg->start, resource_size(cfg));
 		if (IS_ERR(cmn->base))
 			return PTR_ERR(cmn->base);
 		if (cmn->part == PART_CMN600)
-- 
2.43.0

